#VRML_SIM R2023a utf8
  PROTO FourWheelsRobot [
    field SFVec3f    translation  0 0 0
    field SFRotation rotation     0 0 1 0
    field SFFloat    bodyMass     1
    field SFString   controller   "tutorial6"
    field SFString   name   "robot_1"
  ]
  {
    Robot {
      translation IS translation
      rotation IS rotation
      children [
        DistanceSensor {
          translation 0.1 0.04 0
          rotation 0 0 1 0.5235996938995747
          children [
            DEF sensor_shape Shape {
              appearance PBRAppearance {
                baseColor 0 0 1
                metalness 0
              }
              geometry Box {
                size 0.01 0.01 0.01
              }
            }
          ]
          name "ds_left"
          boundingObject USE sensor_shape
          physics Physics {
          }
          recognitionColors [
            0 0 0
          ]
          lookupTable [
            0 0 0
            0.3 30 0
          ]
        }
        DistanceSensor {
          translation 0.1 0 0
          children [
            DEF sensor_shape Shape {
              appearance PBRAppearance {
                baseColor 0 0 1
                metalness 0
              }
              geometry Box {
                size 0.01 0.01 0.01
              }
            }
          ]
          name "ds_center"
          boundingObject USE sensor_shape
          physics Physics {
          }
          lookupTable [
            0 0 0
            0.29 30 0
          ]
        }
        DistanceSensor {
          translation 0.1 -0.04 0
          rotation 0 0 1 -0.5235996938995747
          children [
            DEF sensor_shape Shape {
              appearance PBRAppearance {
                baseColor 0 0 1
                metalness 0
              }
              geometry Box {
                size 0.01 0.01 0.01
              }
            }
          ]
          name "ds_right"
          boundingObject USE sensor_shape
          physics Physics {
          }
          lookupTable [
            0 0 0
            0.3 30 0
          ]
        }
        Solid {
          translation 0 0 0.1
          rotation 0 1 0 0
          children [
            HingeJoint {
              jointParameters HingeJointParameters {
                position 3.2810213838286315e-16
                axis 0 1 0
                anchor 0 -0.01 0.07
              }
              device [
                RotationalMotor {
                  name "motor_camara"
                }
              ]
              endPoint Solid {
                translation 0 -0.01 0.07
                rotation 1 -3.2591849938088094e-10 3.2592407314102743e-10 1.5708014641602872
                children [
                  Camera {
                    rotation 1 0 0 -1.5707996938995747
                    name "CAM"
                    fieldOfView 0.5
                    width 640
                    height 640
                    recognition Recognition {
                      frameThickness 3
                    }
                  }
                  DEF cam_shape Shape {
                    appearance PBRAppearance {
                      baseColor 0.470588 1 0.431373
                      metalness 0
                    }
                    geometry Cylinder {
                      height 0.01
                      radius 0.012
                    }
                  }
                ]
                boundingObject USE cam_shape
                physics Physics {
                }
              }
            }
            DEF cuerpo_camara Shape {
              appearance PBRAppearance {
                baseColor 0.388235 0.784314 0.772549
                metalness 0
              }
              geometry Box {
                size 0.01 0.01 0.15
              }
            }
          ]
          name "estructura_camara"
          boundingObject USE cuerpo_camara
          physics Physics {
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            position 0.5567987504507844
            axis 0 1 0
            anchor 0.05 0.06 0
          }
          device [
            RotationalMotor {
              name "wheel1"
            }
          ]
          endPoint DEF Wheel1 Solid {
            translation 0.05000028913006466 0.06 -1.0320841795982238e-07
            rotation 0.9271412872325822 0.264960822995506 -0.26496187609966576 1.6463775538936514
            children [
              DEF Wheel Shape {
                appearance PBRAppearance {
                  baseColor 0 1 0
                  metalness 0
                }
                geometry Cylinder {
                  height 0.02
                  radius 0.04
                }
              }
            ]
            name "solid wheel1"
            boundingObject USE Wheel
            physics DEF Wheel_PH Physics {
            }
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            position 0.20640154636093624
            axis 0 1 0
            anchor 0.05 -0.06 0
          }
          device [
            RotationalMotor {
              name "wheel2"
            }
          ]
          endPoint DEF Wheel2 Solid {
            translation 0.049999897503821195 -0.059999999566905826 -1.9594232443535167e-06
            rotation -0.9894503176687324 -0.1024402220953058 0.10244056697573102 4.701780087544535
            children [
              USE Wheel
            ]
            name "wheel2"
            boundingObject USE Wheel
            physics USE Wheel_PH
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            position 0.5567984303395221
            axis 0 1 0
            anchor -0.05 0.06 0
          }
          device [
            RotationalMotor {
              name "wheel3"
            }
          ]
          endPoint DEF Wheel3 Solid {
            translation -0.050000767834612436 0.05999999956724689 -1.8054012467920772e-06
            rotation -0.9271413246011201 -0.2649608381104091 0.2649617302266692 4.63680840289176
            children [
              USE Wheel
            ]
            name "wheel3"
            boundingObject USE Wheel
            physics USE Wheel_PH
          }
        }
        HingeJoint {
          jointParameters HingeJointParameters {
            position 0.2064012015620146
            axis 0 1 0
            anchor -0.05 -0.06 0
          }
          device [
            RotationalMotor {
              name "wheel4"
            }
          ]
          endPoint DEF Wheel4 Solid {
            translation -0.050000102481535705 -0.06000000043275322 -1.9594240137107072e-06
            rotation -0.989450352283778 -0.10244005492521074 0.10244039980607017 4.701780122491854
            children [
              USE Wheel
            ]
            name "wheel4"
            boundingObject USE Wheel
            physics USE Wheel_PH
          }
        }
        DEF BODY Shape {
          appearance PBRAppearance {
            baseColor 1 0 0
            metalness 0
            IBLStrength 0
            normalMapFactor 0
            occlusionMapStrength 0
          }
          geometry Box {
            size 0.2 0.1 0.05
          }
        }
      ]
      name IS name
      boundingObject USE BODY
      physics Physics {
        density -1
        mass IS bodyMass
      }
      controller IS controller
    }
  }